File: rules/tempdir.js
const fs = require('fs');
const vscode = require('vscode');

var smell = {

    detect : (fileName, token) => {
        if(token.hasOwnProperty("line")) var lineno = token.line;
        if(token.hasOwnProperty("type")) var tokenType = token.type;
        if(token.hasOwnProperty("name")) var name = token.name;
        if (token.hasOwnProperty("value")) var value = token.value;

        const MSG = 'possible hardcoded temporary directory'
        const WARNING_MSG = MSG+' at line '+ lineno;
        const unwantedDirNames = ['folder', 'directory', 'dir', 'path', 'root', 'tmp', 'temp', 'temporary', 'site', 'log', 'save'];

        if(tokenType == "variable" && name != null && value != null){
            for( const dirName of unwantedDirNames){
                let re = new RegExp(`[_A-Za-z0-9-\.]*${dirName}\\b`);
                if(name.match(re) && smell.isValidPath(value)) smell.triggerAlarm (fileName, MSG, lineno, WARNING_MSG);  
            } 
        }
        else if((tokenType == "list" || tokenType == "set") && name != null) {
            for( const dirName of unwantedDirNames){
                
                let re = new RegExp(`[_A-Za-z0-9-\.]*${dirName}\\b`);
                if(name.match(re)){
                    if(token.hasOwnProperty("values")){
                        for (const value of token.values){
                            if(smell.isValidPath(value)) smell.triggerAlarm (fileName, MSG, lineno, WARNING_MSG);
                        }
                    }
                }
            }           
        }
        else if(tokenType == 'dict' && token.hasOwnProperty('pairs')){
            for(const pair in token.pairs){
                for (const dirName of unwantedDirNames){
                    let re = new RegExp(`[_A-Za-z0-9-\.]*${dirName}\\b`);
                    if(pair[0].match(re) && smell.isValidPath(pair[1])) smell.triggerAlarm (fileName, MSG, lineno, WARNING_MSG);
                }
            }
        }
        else if(tokenType == 'function_call' && token.hasOwnProperty('keywords')){
            for(const keyword in token.keywords){
                for (const dirName of unwantedDirNames){
                    let re = new RegExp(`[_A-Za-z0-9-]*${dirName}\\b`);
                    if(keyword[0].match(re) && smell.isValidPath(keyword[1])) smell.triggerAlarm (fileName, MSG, lineno, WARNING_MSG);
                }
            }
        }
    },

    isValidPath: (path) => {
        // const windowsPathRegex = '^([A-Za-z]?\:?)?\\{1,2}([A-Za-z0-9]*\\)*[A-Za-z0-9]*\\?'
        // const unixPathRegex = '^([A-Za-z0-9]?\:?)?\\{1,2}([A-Za-z0-9]*\\)*[A-Za-z0-9]*\\?'        
        const unixPathRegex = /[_A-Za-z0-9-\/\.\:]*\/{1,2}/
        // const windowsPathRegex = /[_A-Za-z0-9-\/\.\:]*\/{1,2}/
        
        // let unixRegex = new RegExp(unixPathRegex);
        // let windowsRegex = new RegExp(windowsPathRegex);
        
        if (typeof(path) != 'string') return false
        else if(path.match(unixPathRegex)) return true
        else return false
        
        // else if(windowsRegex.test(path)) return true
        
    },
    
    triggerAlarm: (fileName, MSG, lineno, WARNING_MSG) => {
        let backslashSplittedFilePathLength = fileName.split("/").length
        let filenameFromPath = fileName.split("/")[backslashSplittedFilePathLength - 1]
        
        vscode.window.showWarningMessage(MSG +" : "+ filenameFromPath+":"+lineno);
        console.log( "\u001b[1;31m"+"warning: "+MSG +"  location:"+ fileName+":"+lineno);
        fs.appendFileSync(__dirname+'/../warning-logs/project_warnings.csv', fileName+" ,"+WARNING_MSG+"\n");
        // fs.appendFile(__dirname+'/../logs/project_warnings.csv', fileName+","+WARNING_MSG+"\n", (err) => err ? console.log(err): "");
    }
}

module.exports = smell;

File: py-scripts/filters/tempdir.py
import re

class TmpDirectory:
    '''This is the class for detecting hard-coded tmp directories in code'''

    def __init__(self):
        self.warning_message = 'hard-coded tmp directories'
        self.unwanted_dir_names = ['folder', 'directory', 'dir', 'path', 'root', 'tmp', 'temp', 'temporary', 'site', 'log_', 'save', 'upload']                   

    def detect_smell(self, token, src_file):
        try:
            if token.__contains__("line"): lineno = token["line"]
            if token.__contains__("type"): tokenType = token["type"]
            if token.__contains__("name"): name = token["name"]
            if token.__contains__("values"): values = token["values"]
            
            
            if tokenType == "variable" and name is not None and token['value'] is not None:
                for dir_name in self.unwanted_dir_names:
                    if(
                        (re.match(r'[_A-Za-z0-9-]*{dir}\b'.format(dir = dir_name), name.lower().strip()) or re.match(r'\b{dir}[_A-Za-z0-9-]*'.format(dir = dir_name), name.lower().strip())) 
                        and self.is_valid_path(token['value'])
                    ): 
                        self.trigger_alarm(src_file, lineno)
                        break

            elif (tokenType == "list" or tokenType == "set") and token['name'] is not None and token.__contains__('values'):
                smell_found = False

                for dir_name in self.unwanted_dir_names:
                    if(re.match(r'[_A-Za-z0-9-]*{dir}\b'.format(dir = dir_name), name.lower().strip()) or re.match(r'\b{dir}[_A-Za-z0-9-]*'.format(dir = dir_name), name.lower().strip())): 
                        for value in token['values']: 
                            if self.is_valid_path(value):
                                self.trigger_alarm(src_file, lineno)
                                smell_found = True
                                break

                    if smell_found is True: break


            elif tokenType == "dict" and name is not None and token.__contains__('pairs'):
                smell_found = False

                for pair in token['pairs']:
                    for dir_name in self.unwanted_dir_names:
                        if(
                            (re.match(r'[_A-Za-z0-9-]*{dir}\b'.format(dir = dir_name), pair[0].lower().strip()) or re.match(r'\b{dir}[_A-Za-z0-9-]*'.format(dir = dir_name), pair[0].lower().strip()))
                            and self.is_valid_path(pair[1])
                        ):
                                self.trigger_alarm(src_file, lineno)
                                smell_found = True
                                break
                    
                    if smell_found is True: break    

            if tokenType == "function_call" and token.__contains__('keywords'):
                for keyword in token['keywords']:
                    for dir_name in self.unwanted_dir_names:
                        if (
                            len(keyword) == 2 
                            and (re.match(r'[_A-Za-z0-9-]*{dir}\b'.format(dir = dir_name), keyword[0].lower().strip()) or re.match(r'\b{dir}[_A-Za-z0-9-]*'.format(dir = dir_name), keyword[0].lower().strip())) 
                            and self.is_valid_path(keyword[1])
                        ): 
                            self.trigger_alarm(src_file, lineno)
                            break

        except Exception as error: print(str(error))

    def is_valid_path(self, value):

        if value is None: return False
        if isinstance(value, str) is False: return False
        if len(value) == 0: return False
        if value == '/': return False
        if value == '//': return False
        if value == '\\': return False

        unix_path_reg = r'^(~?((\.{1,2}|\/?)*[a-zA-Z0-9]*\/)+)[a-zA-Z0-9]*\/?'
        windows_path_reg = r'^([A-Za-z]?\:?)?\\{1,2}([A-Za-z0-9]*\\)*[A-Za-z0-9]*\\?'

        if re.fullmatch(unix_path_reg, value): return True
        elif re.fullmatch(windows_path_reg, value): return True
        else: return False
    
    def trigger_alarm(self, src_file, lineno):
        print(src_file +":"+ str(lineno)+" ,"+self.warning_message)
    